components/menu/src/menu/use-menu.js

```diff
@@ -0,0 +1,91 @@
+import { useRef, useState, useEffect, useCallback } from 'react'
+import { getFocusableItemsIndices } from './helpers.js'
+
+export const useMenuNavigation = (children) => {
+    const menuRef = useRef(null)
+    const [focusableItemsIndices, setFocusableItemsIndices] = useState(null)
+    const [activeItemIndex, setActiveItemIndex] = useState(-1)
+
+    // Initializes the indices for focusable items
+    useEffect(() => {
+        if (menuRef) {
+            const menuItems = Array.from(menuRef.current.children)
+            const itemsIndices = getFocusableItemsIndices(menuItems)
+            setFocusableItemsIndices(itemsIndices)
+        }
+    }, [children])
+
+    // Focus the active menu child
+    useEffect(() => {
+        if (menuRef) {
+            if (focusableItemsIndices && activeItemIndex > -1) {
+                const currentIndex = focusableItemsIndices[activeItemIndex]
+                menuRef.current.children[currentIndex].focus()
+            }
+        }
+    }, [activeItemIndex, focusableItemsIndices])
+
+    // Navigate through focusable children using arrow keys
+    // Trigger actionable items
+    const handleKeyDown = useCallback(
+        (event) => {
+            const totalFocusablePositions = focusableItemsIndices?.length
+            const lastIndex = totalFocusablePositions - 1
+            switch (event.key) {
+                case 'ArrowUp':
+                    event.preventDefault()
+                    setActiveItemIndex(
+                        activeItemIndex > 0 ? activeItemIndex - 1 : lastIndex
+                    )
+                    break
+                case 'ArrowDown':
+                    event.preventDefault()
+                    setActiveItemIndex(
+                        activeItemIndex >= lastIndex ? 0 : activeItemIndex + 1
+                    )
+                    break
+                case 'Enter':
+                case ' ':
+                    event.preventDefault()
+                    if (event.target.nodeName === 'LI') {
+                        event.target.children[0].click()
+                    }
+                    break
+                default:
+                    break
+            }
+        },
+        [activeItemIndex, focusableItemsIndices?.length]
+    )
+
+    // Event listeners for menu focus and key handling
+    useEffect(() => {
+        if (!menuRef) {
+            return
+        }
+
+        const menu = menuRef.current
+
+        // Focus the first menu item when the menu receives focus
+        const handleFocus = (event) => {
+            if (event.target === menuRef.current) {
+                const firstItemIndex = focusableItemsIndices?.[0]
+                menuRef.current.children[firstItemIndex].focus()
```

Comment:
this bit errors in the storybook for AutoHideFirstSectionHeaderDivider. it is because firstItemIndex in this case is undefined. Maybe we want to wrap this in a conditional?
```suggestion
                menuRef.current.children[firstItemIndex].focus()
```

Reply:
Okay, so the AutoHideFirstSectionHeaderDivider only contains 2 section headers as its children and in my implementation, those do not receive visual focus. So i have to adjust for cases where the menu does not contain any focusable items. Good catch, thank you